20210112_map/src/de/telran/OurHashMap.java

```diff
@@ -0,0 +1,187 @@
+package de.telran;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+/**
+ * The implementation of OurMap cannot contain null key
+ * @param <K>
+ * @param <V>
+ */
+
+public class OurHashMap<K, V> implements OurMap<K, V> {
+    private static final double DEFAULT_LOAD_FACTOR = 0.75;
+    private static final int INITIAL_CAPACITY = 16;
+
+    private Pair<K, V>[] source;
+    private int size;
+    private double loadFactor;
+    private int capacity;
+
+    //Пустой конструктор
+    public OurHashMap(){
+        source = new Pair[INITIAL_CAPACITY];
+        capacity = INITIAL_CAPACITY;
+        size = 0;
+        loadFactor = DEFAULT_LOAD_FACTOR;
+    }
+
+    //Это необязательный метод, мы его написали как донастройку hashCode
+    static int hash(Object key){
+        return Math.abs(key.hashCode());//эта функция не очень хорошая, так как все значения этой функции
+                            // могут находиться в довольно низком диапазоне
+    }
+
+    public OurHashMap(double loadFactor) {
+        this();//Конструкторы можно вызывать друг из друга
+        this.loadFactor = loadFactor;
+    }
+
+    @Override
+    public V put(K key, V value) {
+        if (size >= loadFactor * capacity){
+            resize();
+        }
+        Pair<K, V> pair = find(key);
+
+        if (pair != null) {
+            //V res = pair.value;//Берем старое значение
+            pair.value = value;
+            //size++;
+            //return res;
+            return pair.value;
+        }
+
+        int index = hash(key) % capacity;
+        Pair<K, V> newPair = new Pair<>(key, value, source[index]);
+        source[index] = newPair;
+        size++;
+        return newPair.value;
+    }
+
+    private void resize(){
+        int newCapacity = capacity * 2;
+        Pair<K, V>[] newSource = new Pair[newCapacity];
+
+        for (Pair<K, V> pair : source) {
+            if (pair != null) {
+                int index = hash(pair.key) % newCapacity;
+                newSource[index] = pair;
+            }
+        }
+        source = newSource;
+    }
+
+    private Pair<K, V> find(K key){
+        //int index = hash(key) % capacity;
+        int index = Math.abs(key.hashCode() % capacity);
+        Pair<K, V> current = source[index];
+        while (current != null){
+            if (key.equals(current.key)){
+                return current;
+            } else {
+                current = current.next;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public V get(K key) {
+        if (source.length == 0) {
+            throw new NullPointerException();
+        }
+        Pair<K, V> pair = find(key);
+        if (pair == null) {
+            throw new NullPointerException();
```

Comment:
the contract of the method says that if there is no pair with this key, return null as a result.